src/experimental/tools/download2go/index.ts

```diff
@@ -0,0 +1,415 @@
+        ? "d2g-canalplus"
+        : options.nameDB;
+    this.activeDownloads = {};
+    this.activePauses = {};
+    this.db = null;
+    this.emitter = {
+      trigger: (eventName, payload) => this.trigger(eventName, payload),
+    };
+  }
+
+  /**
+   * Initialize an IndexDB instance.
+   * @returns {Promise<void>}
+   */
+  initDB(): Promise<IDBPDatabase> {
+    return new Promise((resolve, reject) => {
+      setUpDb(this.nameDB)
+        .then(db => {
+          this.db = db;
+          resolve(db);
+        })
+        .catch((error: Error) => {
+          this.trigger("error", { action: "initDB", error });
+          reject(error);
+        });
+    });
+  }
+
+  /**
+   * Start a download from scratch.
+   * @param {Object<ISettingsDownloader>} settings
+   * @returns {Promise.<void>}
+   */
+  async download(options: IInitSettings): Promise<void> {
+    try {
+      const db = this.db;
+      if (db === null) {
+        throw new Error("The IndexDB database has not been created!");
+      }
+      await checkInitDownloaderOptions(options, db, this.activeDownloads);
+      const { metaData, contentID, transport } = options;
+      const pause$ = new AsyncSubject<void>();
+      this.activePauses[contentID] = pause$;
+      const downloadManager = new DownloadManager({
+        emitter: this.emitter,
+        db,
+      });
+      const initDownloadSub = downloadManager
+        .initDownload(options, pause$)
+        .subscribe(
+          ([download]) => {
+            if (download === null) {
+              return;
+            }
+            const { progress, manifest, audio, video, text, size } = download;
+            if (manifest === null) {
+              return;
+            }
+            db.put("manifests", {
+              contentID,
+              transport,
+              manifest,
+              builder: { video, audio, text },
+              progress,
+              size,
+              duration: manifest.getDuration(),
+              metaData,
+            }).then(() =>
+              this.trigger("insertDB", {
+                action: "store",
+                contentID,
+                progress: progress.percentage,
+              })
+            ).catch((err: Error) => {
+              this.trigger("error", {
+                action: "resume-downloader",
+                error: err,
+                contentID,
+              });
+            });
+          },
+          (error: Error) => this.trigger("error", { action: "init-downloader", error })
+        );
+      this.activeDownloads[contentID] = initDownloadSub;
+    } catch (error) {
+      const err = error as Error;
+      this.trigger("error", {
+        action: "init-downloader",
+        error: err,
+        contentID: options.contentID,
+      });
+    }
+  }
+
+  /**
+   * Resume a download already started earlier.
+   * @param {string} contentID
+   * @returns {Promise.<void>}
+   */
+  async resume(contentID: string): Promise<void> {
```

Comment:
I'm using Promise here because I want to know when the `resume` action is done.

Because what if the user `pause` and `resume` too quickly, we will have no time set up everything and can lead to a race condition or blocked process.